lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -235,4 +235,35 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
+# Collect text output into JSON. save_results() is called too early for this.
+sub finalize_results {
+    my ($self) = @_;
+
+    my $dir = $self->job->result_dir();
+    return unless $dir;
+    my $file = path($dir, "details-" . $self->name . ".json");
+    return unless -e $file;
+    my $results;
+
+    try {
+        $results = decode_json($file->slurp);
+    }
+    catch {
+        return;
+    };
+
+    $results = {details => $results} if ref($results) eq 'ARRAY';
+
+    for my $step (@{$results->{details}}) {
+        next unless $step->{text};
+        my $txtfile = path($dir, $step->{text});
+        next unless -e $txtfile;
+        $step->{text_data} = $txtfile->slurp;
+    }
+
+    my $tmpfile = $file->sibling($file->basename . '.tmp');
+    $tmpfile->spurt(encode_json($results));
+    rename $tmpfile->to_string, $file->to_string;
```

Comment:
I don't think we use autodie here so possible errors of `rename` should be handled.